core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -940,11 +941,24 @@ void reregister(
       assert equivRel.getClass() == rel.getClass();
       assert equivRel.getTraitSet().equals(rel.getTraitSet());
 
+      mergeIdenticalRels(equivRel, rel);
       return;
     }
 
     // Add the relational expression into the correct set and subset.
-    addRelToSet(rel, set);
+    if (!prunedNodes.contains(rel)) {
+      addRelToSet(rel, set);
+    }
+  }
+
+  /**
+   * Update relNode related information when two identical relNodes are
+   * de-duped.
+   */
+  private void mergeIdenticalRels(RelNode to, RelNode from) {
```

Comment:
from, to is confusing. Can we use equivRel, rel instead? And the comment doesn't elaborate what does it do.

Reply:
I changed it to RelNode rel, RelNode duplicate, where duplicate is the one that's about to be discarded.